### PR TITLE
health: Create package to determine health

### PR DESCRIPTION
**Pull request by gvso**
This creates a new health package that has a main IsHealthy function to check if the metrics for the candidate revision meet the configuration criteria.

### CODE REVIEW

File: internal/metrics/metrics.go [hunk 1]
```diff
@@ -32,3 +32,19 @@ type Metrics interface {
 	Latency(ctx context.Context, query Query, offset time.Duration, alignReduceType AlignReduce) (float64, error)
 	ErrorRate(ctx context.Context, query Query, offset time.Duration) (float64, error)
 }
+
+// PercentileToAlignReduce takes a percentile value maps it to a AlignReduce
+// value.
+//
+// TODO: once we start supporting any percentile value, this should not be
```
Comment (ahmetb) [inline internal/metrics/metrics.go:39]: why are we not able to support floats right away?

Comment (gvso) [reply internal/metrics/metrics.go:39]: Since we decided to use the REST API instead of MQL first, we don't need (and can't) support floats. The REST API, only allowed specific percentiles (99, 95, 50)

Comment (ahmetb) [reply internal/metrics/metrics.go:39]: makes sense. let's have validation that if values like 97 is specified it should return error instead of defaulting to some unwanted value.

Comment (gvso) [reply internal/metrics/metrics.go:40]: Done. I'm using `fmt.Errorf` since `pkg/errors` doesn't have a constructor with formatting.

Comment (ahmetb) [reply internal/metrics/metrics.go:40]: don't use fmt.Errorf. Try errors.Errorf.

Comment (gvso) [reply internal/metrics/metrics.go:40]: Will do so. That actually makes more sense. It's interesting that go-lint recommended `fmt.Errorf`

File: pkg/health/health.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+)
+
+// Checker validates that revision is healthy.
+type Checker interface {
+	IsHealthy(ctx context.Context, query metrics.Query, healthCheckOffset time.Duration, metricsChecks []config.Metric) (bool, error)
+}
+
+// Health is the candidate's health status checker.
+type Health struct {
+	Metrics metrics.Metrics
```
Comment (ahmetb) [inline pkg/health/health.go:19]: unexport

File: pkg/health/health.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+// New initializes a health checker.
+func New(metrics metrics.Metrics) *Health {
+	return &Health{
+		Metrics: metrics,
+	}
+}
+
+// IsHealthy checks if a revision is healthy based on the metrics configuration.
+func (h *Health) IsHealthy(ctx context.Context, query metrics.Query, healthCheckOffset time.Duration, metricsChecks []config.Metric) (bool, error) {
```
Comment (ahmetb) [inline pkg/health/health.go:30]: 1. Health doesn't have to be a type. You can pass `metrics.Metrics` here as well.
2. IsHealthy doesn't have to be defined on a type. see 1
3. You can easily test IsHealthy by passing mocks, no need for `interface Checker`

File: pkg/health/health.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	for _, check := range metricsChecks {
+		var err error
+		switch check.Type {
+		case config.LatencyMetricsCheck:
+			healthy, err = h.checkLatency(ctx, query, healthCheckOffset, check.Percentile, check.Max)
+			break
+		case config.ErrorRateMetricsCheck:
+			healthy, err = h.checkErrorRate(ctx, query, healthCheckOffset, check.Max)
+			break
+		}
```
Comment (ahmetb) [inline pkg/health/health.go:41]: default: return err unimplemented %s

File: pkg/health/health.go (hunk 3 above)
Comment (ahmetb) [inline pkg/health/health.go:30]: Maybe we need to do this in another PR:

If you just return `bool` for fail/pass, we cannot capture details like "what metrics did we see", "why did we make this decision". 

As part of our design we said we'll provide these insights back to the users. So probably we need to export something like []CheckResults that explains result of each criteria checked and if it was "met" or "not".

File: internal/metrics/metrics.go [hunk 5]
```diff
@@ -3,6 +3,8 @@ package metrics
 import (
```
Comment (gvso) [inline internal/metrics/metrics.go:3]: Ignore this. Same as #17

File: internal/metrics/metrics_test.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
+package metrics_test
```
Comment (gvso) [inline internal/metrics/metrics_test.go:1]: Ignore this. Same as #17

File: pkg/health/health.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+// Diagnose attempts to determine the health of a revision.
+//
+// If the minimum number of requests is not met, then health cannot be
+// determined and Diagnosis.EnoughRequests is set to false.
+//
+// Otherwise, all metrics criteria are checked to determine if the revision is
+// healthy.
+func Diagnose(ctx context.Context, provider metrics.Metrics, query metrics.Query,
```
Comment (gvso) [inline pkg/health/health.go:37]: This does not check request count yet since it depends on #13 . I will do it in a following PR

File: pkg/health/health.go [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+	IsHealthy          bool
+}
+
+// CheckResult is information about a metrics criteria check.
+type CheckResult struct {
+	MetricsType   config.MetricsCheck
+	MinOrMaxValue interface{}
+	ActualValue   interface{}
+	IsCriteriaMet bool
+	Reason        string
```
Comment (ahmetb) [inline pkg/health/health.go:27]: it seems like MinOrMaxValue (rename: Threshold or ExpectedValue) and ActualValue are what we need in this type.

IsCriteriaMet is ok to.

but the type etc can be inferred from the []config.Metric we passed in.

Comment (ahmetb) [reply pkg/health/health.go:27]: also just use float64

File: pkg/health/health.go [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+	metricsValues, err := CollectMetrics(ctx, provider, query, offset, healthCriteria)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not collect metrics")
+	}
+
+	isHealthy := true
+	var results, failedResults []*CheckResult
+	for i, criteria := range healthCriteria {
+		result := determineResult(criteria.Type, criteria.Min, criteria.Max, metricsValues[i])
```
Comment (ahmetb) [inline pkg/health/health.go:48]: update config.Metric to have Threshold (not Min / Max) and here in this method we know if it should be used as ≤ or ≥

File: pkg/health/health.go [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+		if !result.IsCriteriaMet {
+			isHealthy = false
+			failedResults = append(failedResults, result)
+		}
+	}
+
+	return &Diagnosis{
+		EnoughRequests:     true,
+		CheckResults:       results,
+		FailedCheckResults: failedResults,
```
Comment (ahmetb) [inline pkg/health/health.go:60]: no need for this

File: pkg/health/health.go [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+		result := determineResult(criteria.Type, criteria.Min, criteria.Max, metricsValues[i])
+		results = append(results, result)
+
+		if !result.IsCriteriaMet {
+			isHealthy = false
+			failedResults = append(failedResults, result)
+		}
+	}
+
+	return &Diagnosis{
```
Comment (ahmetb) [inline pkg/health/health.go:57]: ideally this structure should also represent that no rollout operation should be done (halt/noop) based on minrequests.

similar to IsHealthy but different.

this is probably an enum DiagnosisResult { UNKNOWN=0 , INCONCLUSIVE, HEALTHY= , UNHEALTHY }

File: pkg/health/health.go [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+		FailedCheckResults: failedResults,
+		IsHealthy:          isHealthy,
+	}, nil
+}
+
+// CollectMetrics returns an array of values collected for each of the specified
+// metrics criteria.
+func CollectMetrics(ctx context.Context, provider metrics.Metrics, query metrics.Query, offset time.Duration, healthCriteria []config.Metric) ([]interface{}, error) {
+
+	var values []interface{}
```
Comment (ahmetb) [inline pkg/health/health.go:69]: don't use empty interface

File: pkg/health/health.go [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+	for _, criteria := range healthCriteria {
+		var value interface{}
+		var err error
+
+		switch criteria.Type {
+		case config.LatencyMetricsCheck:
+			value, err = latency(ctx, provider, query, offset, criteria.Percentile)
+			break
+		case config.ErrorRateMetricsCheck:
+			value, err = errorRate(ctx, provider, query, offset)
```
Comment (ahmetb) [inline pkg/health/health.go:79]: add logs

File: pkg/health/health.go [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/pkg/config"
+	"github.com/pkg/errors"
+)
+
+// Diagnosis is the information about the health of the revision.
+type Diagnosis struct {
+	EnoughRequests     bool
+	CheckResults       []*CheckResult
+	FailedCheckResults []*CheckResult
+	IsHealthy          bool
```
Comment (ahmetb) [inline pkg/health/health.go:18]: ideally this structure should also represent that no rollout operation should be done (halt/noop) based on minrequests.

similar to IsHealthy but different.

this is probably an enum DiagnosisResult { UNKNOWN=0 , INCONCLUSIVE, HEALTHY= , UNHEALTHY }

File: pkg/health/health.go [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+	latency, err := provider.Latency(ctx, query, offset, alignerReducer)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to get latency metrics")
+	}
+
+	return latency, nil
+}
+
+// errorRate returns the percentage of errors during the given offset.
+func errorRate(ctx context.Context, provider metrics.Metrics, query metrics.Query, offset time.Duration) (float64, error) {
```
Comment (ahmetb) [inline pkg/health/health.go:150]: errorRatePercent

File: pkg/health/health.go (hunk 13 above)
Comment (gvso) [reply pkg/health/health.go]: This will be done in a following PR after we integrate this package with the `rollout`

File: pkg/health/health.go (hunk 9 above)
Comment (gvso) [reply pkg/health/health.go]: Done at #20

File: pkg/health/health.go [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// DiagnosisResult is a possible result after a diagnosis.
+type DiagnosisResult int32
+
+// Possible diagnosis results.
+const (
+	Unknown      DiagnosisResult = 0
```
Comment (ahmetb) [inline pkg/health/health.go:19]: use iota syntax.

File: pkg/health/health.go [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/internal/metrics"
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/internal/util"
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/pkg/config"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// DiagnosisResult is a possible result after a diagnosis.
+type DiagnosisResult int32
```
Comment (ahmetb) [inline pkg/health/health.go:15]: why int32? can't this just be int

File: pkg/health/health.go [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+	Unknown      DiagnosisResult = 0
+	Inconclusive DiagnosisResult = 1
+	Healthy      DiagnosisResult = 2
+	Unhealthy    DiagnosisResult = 3
+)
+
+// Diagnosis is the information about the health of the revision.
+type Diagnosis struct {
+	OverallResult DiagnosisResult
+	CheckResults  []*CheckResult
```
Comment (ahmetb) [inline pkg/health/health.go:28]: why is this pointer? use pointers very sparingly

File: pkg/health/health.go [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+func Diagnose(ctx context.Context, provider metrics.Metrics, query metrics.Query,
+	offset time.Duration, minRequests int64, healthCriteria []config.Metric) (*Diagnosis, error) {
+
+	logger := util.LoggerFromContext(ctx)
+	metricsValues, err := CollectMetrics(ctx, provider, query, offset, healthCriteria)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not collect metrics")
+	}
+
+	overallResult := Healthy
```
Comment (ahmetb) [inline pkg/health/health.go:54]: this is potentially a problem if no healthCriteria is specified and you return `Healthy`. that's why we define `Inconclusive` and `Unknown` for.

if you say `var overallResult DiagnosisResult` it will default to `Unknown`.

Comment (ahmetb) [reply pkg/health/health.go:54]: to be clear my concern is:  It's not defensive enough for a critical task such as rollout. user is much better off if we don't roll out or rollback, rather than blindly rolling forward due to a bug in our tool.

I need you to approach to the problem with this mindset.

File: pkg/health/health.go [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+			overallResult = Unhealthy
+
+			logger := logger.WithFields(logrus.Fields{
+				"metricsType": criteria.Type,
+				"threshold":   criteria.Threshold,
+				"actualValue": result.ActualValue,
+			})
+			// If criteria is a latency, we want to log the percentile as well
+			if criteria.Type == config.LatencyMetricsCheck {
+				logger = logger.WithField("percentile", criteria.Percentile)
```
Comment (ahmetb) [inline pkg/health/health.go:70]: just log it always, it'll be a percentile=0 field that doesn't mean anything. it's cleaner that way (+log stms align in human-readable output)

File: pkg/health/health.go [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+		return nil, errors.Wrap(err, "could not collect metrics")
+	}
+
+	overallResult := Healthy
+	var results []*CheckResult
+	for i, criteria := range healthCriteria {
+		result := determineResult(criteria.Type, criteria.Threshold, metricsValues[i])
+		results = append(results, result)
+
+		if !result.IsCriteriaMet {
```
Comment (ahmetb) [inline pkg/health/health.go:60]: I probably explained "avoid if-nesting" if you can more than a handful times now. :) 

What if you said:

```
for healthCriteria {

  if result.IsCriteriaMet {
    continue
  }

  // things you do if unhealthy
  // and you don't need to indent it
}
```

Comment (ahmetb) [reply pkg/health/health.go:60]: any time you write a lengthy `if { ... }` block I need you to keep this in mind: can you move the block outside the `if` somehow or not

File: pkg/health/health.go [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+			logger := logger.WithFields(logrus.Fields{
+				"metricsType": criteria.Type,
+				"threshold":   criteria.Threshold,
+				"actualValue": result.ActualValue,
+			})
+			// If criteria is a latency, we want to log the percentile as well
+			if criteria.Type == config.LatencyMetricsCheck {
+				logger = logger.WithField("percentile", criteria.Percentile)
+			}
+			logger.Debug("criteria was not met")
```
Comment (ahmetb) [inline pkg/health/health.go:72]: why don't you initialize logger outside the if and also log `criteria met`

**Comment by ahmetb** [comment]
cc: @grayside this is probably the most important PR of the project. If you have time, PTAL.

In this PR we create some core machinery to look at a set of metrics thresholds and look at Stackdriver Monitoring API and make a decision (roll forward/back, noop).

File: pkg/health/health_test.go [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+func TestDiagnose(t *testing.T) {
+	metricsMock := &metricsMocker.Metrics{}
+	metricsMock.LatencyFn = func(ctx context.Context, query metrics.Query, offset time.Duration, alignReduceType metrics.AlignReduce) (float64, error) {
+		return 500, nil
+	}
+	metricsMock.ErrorRateFn = func(ctx context.Context, query metrics.Query, offset time.Duration) (float64, error) {
+		return 0.01, nil
+	}
+
+	tests := []struct {
```
Comment (grayside) [inline pkg/health/health_test.go:24]: It might be good to test metrics and thresholds set to 0. Even if that is working fine now, that seems likely to raise unexpected behavior in any future refactoring.

File: pkg/health/health.go [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+// was met or not.
+func determineResult(metricsType config.MetricsCheck, threshold float64, actualValue float64) *CheckResult {
+	result := &CheckResult{ActualValue: actualValue, Threshold: threshold}
+
+	// As of now, the supported health criteria (latency and error rate) need to
+	// be less than the threshold. So, this is sufficient for now but might need
+	// to change to a switch statement when criteria with a minimum threshold is
+	// added.
+	if actualValue <= threshold {
+		result.IsCriteriaMet = true
```
[the remark this answers is not in the document]
Comment (ahmetb) [reply pkg/health/health.go:124]: @grayside so it's actually hard to look at `CheckResult` object and tell if it's a positive result or negative. 

e.g. for latency metrics, threshold=100 actual=200 → requirement is **failed**
however, for "min request" metrics threshold=100 actual=200 → requirement is **met**

the method that generates this CheckResult decides if those numbers mean something good or bad.

in case of "min request" metrics is **not met** → we simply do noop (wait more requests to happen)

however in case of "latency" metrics **not met** → means "rollback".

File: pkg/health/health.go [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+}
+
+// Diagnose attempts to determine the health of a revision.
+//
+// If the minimum number of requests is not met, then health cannot be
+// determined and Diagnosis.EnoughRequests is set to false.
+//
+// Otherwise, all metrics criteria are checked to determine if the revision is
+// healthy.
+func Diagnose(ctx context.Context, provider metrics.Metrics, query metrics.Query,
```
[the remark this answers is not in the document]
Comment (ahmetb) [reply pkg/health/health.go:45]: I’ll try to explain the context here:

1. `provider` is a metrics backend that lets you query certain type of metrics. However, we specify "filters" such as "service_name=X,revision=Y,project=Z", and those (under the covers) might be implemented differently. 

    though I agree, actually we could make that as part of initialization process to the MetricsProvider (and initialize a different `provider` for each svc/revision/project/...)

2. `offset` is for being able to specify "metrics for last X minutes" (for example Stackdriver actually doesn't support last X minutes, rather, we need to compute startDate/endDate inside its metric provider impl).

3. `minRequests` is something we will (over time) transition to an actual `config.Metric` criteria (just like latency) so ideally this will go away from this method signature.

File: pkg/health/health.go [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+		result.IsCriteriaMet = true
+	}
+	return result
+}
+
+// latency returns the latency for the given offset and percentile.
+func latency(ctx context.Context, provider metrics.Metrics, query metrics.Query, offset time.Duration, percentile float64) (float64, error) {
+	alignerReducer, err := metrics.PercentileToAlignReduce(percentile)
+	if err != nil {
+		return 0, errors.Wrap(err, "invalid percentile")
```
[the remark this answers is not in the document]
Comment (ahmetb) [reply pkg/health/health.go:133]: we get the actual percentile value from the user (e.g. p94) however the API might not support that. so if it's an unsupported value, we return a parse error here.

[the remark this answers is not in the document]
Comment (ahmetb) [reply pkg/health/health.go:133]: however I agree we should probably notice this "parse error" waaay early on, before even starting to diagnose the service.

File: pkg/health/health.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (grayside) [reply pkg/health/health.go:124]: Ah, that makes sense, and reading again I see the comment about the forthcoming switch statement. That seems like it will be a bit complex. I can't help but go to the idea that there should be an interface such as the following (reusing the same verbs being discussed above, but more locally this time):

```go
type Metric interface {
  Assess(context.Context, metrics.Metrics, time.Duration, []config.Metrics[]) (*CheckResult, error)
  Diagnose() bool
}
```

Then the individual Metrics can be implemented with a bit more isolation and testability.
The current structure seems like it will work okay for 2-3 metrics but start to become less manageable thereafter.

File: pkg/health/health.go (hunk 19 above)
Comment (gvso) [reply pkg/health/health.go:54]: Thanks! My initial thought was that `Inconclusive` is the value returned when there are no enough requests, but that's still not implemented.

I was not sure if the best approach to empty criteria was roll forward (maybe the user didn't want health checks) or do nothing (`Unknown`/`Inconclusive`).

File: pkg/health/health.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (gvso) [reply pkg/health/health.go:124]: Thanks @grayside! I think I understand your concern, but I am a little bit confused about your expected interface for the caller of this interface.

The idea is that you get a `CheckResult` per metric criteria since the `CheckResult` contains the expected threshold, the actual value, and whether the criteria is met based on those previous two values.

1) Do you want to pass a single `config.Metric` and get a single `CheckResult`?
```
for criteria := range healthCriteria {
    results := append(results, Assess(ctx, metricsProvider, ....)) // ignore error
}
```

2) Do you want to pass an array of metrics criteria an get an array of `CheckResult`?

3) Something else?

Also, I don't see how this will prevent the `switch` when determining results since using comparing `actualValue <= threshold` or `threshold <= actualValue` depends on the metrics criteria type (unless we pass a comparison function, which sounds like an overkill here).

File: pkg/health/health.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (gvso) [reply pkg/health/health.go:133]: The config checks are done in `main`, though checks for health criteria has not been added yet (#27). This is done here as a defensive approach as we have done with many other things. I would keep the `float64` here since we would need it once we support any percentile value.

File: pkg/health/health.go [hunk 27]
```diff
@@ -0,0 +1,160 @@
...
+			// If criteria is a latency, we want to log the percentile as well
+			if criteria.Type == config.LatencyMetricsCheck {
+				logger = logger.WithField("percentile", criteria.Percentile)
+			}
+			logger.Debug("criteria was not met")
+		}
+	}
+
+	return &Diagnosis{
+		OverallResult: overallResult,
```
[the remark this answers is not in the document]
Comment (gvso) [reply pkg/health/health.go:77]: I like this approach, what about doing this:

```
results := health.Assess(ctx, provider, query, offset, healthCriteria) // returns []health.CheckResult
diagnosis := health.Diagnose(healthCriteria, results)
```

The above approach works too, it just means an extra struct containing `[]CheckResult`

File: pkg/health/health.go (hunk 16 above)
Comment (gvso) [reply pkg/health/health.go:19]: Nice! I didn't know about this.

File: pkg/health/health.go (hunk 19 above)
Comment (ahmetb) [reply pkg/health/health.go:54]: > I was not sure if the best approach to empty criteria was roll forward

Right, it's not a good approach.

Inconclusive or Unknown is better. But best might be returning error if no criteria is specified (terminal error).

File: pkg/health/health.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (ahmetb) [reply pkg/health/health.go:124]: I'm a bit +1 on this being an overkill right now.
Realistically we'll only have `a≤b` or `b≤a` comparisons as far as metrics go.
So creating a `LatencyMetric` or `RequestCountMetric` types probably are overkill.

Though I agree, Adam’s approach long-term is definitely appealing. I'd say no need to refactor this right now.

File: pkg/health/health.go [hunk 28]
```diff
@@ -0,0 +1,160 @@
...
+		CheckResults:  results,
+	}, nil
+}
+
+// CollectMetrics returns an array of values collected for each of the specified
+// metrics criteria.
+func CollectMetrics(ctx context.Context, provider metrics.Metrics, query metrics.Query, offset time.Duration, healthCriteria []config.Metric) ([]float64, error) {
+	logger := util.LoggerFromContext(ctx)
+	logger.Debug("start collecting metrics")
```
[the remark this answers is not in the document]
Comment (ahmetb) [reply pkg/health/health.go:86]: I don't know how OT would play into here. Are you talking about "querying OT" or "using OT to trace the operator tool itself"?

We would look metrics storages (Stackdriver API, Prometheus) and those don't know anything about OT when it comes to querying.

File: pkg/health/health.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (ahmetb) [reply pkg/health/health.go:77]: Yeah, SGTM. Please capture this in an issue and let's do it another PR.

**Comment by ahmetb** [comment]
Let’s address the quick actionable feedback and open issues for outstanding long-term issues and send follow-up PRs. This CL is getting pretty beefy.

File: pkg/health/health.go (hunk 21 above)
Comment (gvso) [reply pkg/health/health.go:60]: My bad. This was a simple block before the last change. I will take this into account when making changes too

File: pkg/health/health.go (hunk 22 above)
Comment (gvso) [reply pkg/health/health.go]: Fixed

File: pkg/health/health.go (hunk 21 above)
Comment (gvso) [reply pkg/health/health.go]: Fixed!

File: pkg/health/health.go (hunk 20 above)
Comment (gvso) [reply pkg/health/health.go]: Fixed!

File: pkg/health/health.go (hunk 19 above)
Comment (gvso) [reply pkg/health/health.go:54]: Let's address this at #30

File: pkg/health/health.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (gvso) [reply pkg/health/health.go:76]: Keeping track of this at #30

File: pkg/health/health_test.go (hunk 23 above)
Comment (gvso) [reply pkg/health/health_test.go:24]: Keeping track at #31

**Comment by gvso** [comment]
I have addressed the quick actionable feedback, and I am keeping track of the other issues at #30 and #31
